torch/testing/_internal/common_nn.py

```diff
@@ -4881,103 +4948,103 @@ def _do_test(self, test_case, module, input):
                     test_case.assertIsInstance(p, torch.cuda.FloatTensor)
                     test_case.assertEqual(p.get_device(), 0)
 
                 # to CPU
-                input = input.cpu()
+                input_tuple = tuple(t.cpu() for t in input_tuple)
                 module.cpu()
-                module(input)
+                module(*input_tuple)
                 for p in module.parameters():
                     test_case.assertIsInstance(p, torch.FloatTensor)
 
                 # back to GPU0
-                input = input.cuda()
+                input_tuple = tuple(t.cuda() for t in input_tuple)
                 module.cuda()
-                module(input)
+                module(*input_tuple)
                 for p in module.parameters():
                     test_case.assertIsInstance(p, torch.cuda.FloatTensor)
                     test_case.assertEqual(p.get_device(), 0)
 
                 # test that forwards of module runs correctly without cuDNN
                 if self.cudnn:
                     with torch.backends.cudnn.flags(enabled=False):
-                        module(input)
+                        module(*input_tuple)
                         for p in module.parameters():
                             test_case.assertIsInstance(p, torch.cuda.FloatTensor)
                             test_case.assertEqual(p.get_device(), 0)
 
                 if torch.cuda.device_count() >= 2:
                     # test cross-GPU transfer works
                     # to GPU1
-                    input = input.cuda(1)
+                    input_tuple = tuple(t.cuda(1) for t in input_tuple)
                     module.cuda(1)
                     with torch.cuda.device(1):
-                        module(input)
+                        module(*input_tuple)
                     for p in module.parameters():
                         test_case.assertIsInstance(p, torch.cuda.FloatTensor)
                         test_case.assertEqual(p.get_device(), 1)
 
                 if not self.skip_double:
                     # test double()
-                    input = input.double().cuda()
+                    input_tuple = tuple(t.double().cuda() for t in input_tuple)
                     module.double().cuda()
-                    module(input)
+                    module(*input_tuple)
                     for p in module.parameters():
                         test_case.assertIsInstance(p, torch.cuda.DoubleTensor)
                         test_case.assertEqual(p.get_device(), 0)
 
                 # test half()
-                input = input.half().cuda()
+                input_tuple = tuple(t.half().cuda() for t in input_tuple)
                 module.half().cuda()
-                module(input)
+                module(*input_tuple)
                 for p in module.parameters():
                     test_case.assertIsInstance(p, torch.cuda.HalfTensor)
                     test_case.assertEqual(p.get_device(), 0)
```

Comment:
nit: There's a lot of repetition here in the device-swapping tests. It's a pre-existing problem but it would be nice if we could clean it up, since we're modifying the code anyways. It took me a while to understand what was being tested here.

Reply:
@zou3519 
Ah, yes, I understand this. I can consolidate this into one function and call it with different type and device id. 

When I wrote this part, I was debating on this. If written them into one function, then we loose the flexibility to change one of them if needed. And on the other hand, some of them test both assertIsInstance and assertEqual, some of them test assertIsInstance only. So if you consolidate them into one function, you will have some kind of flag indicate whether to test assertEqual, and the function call will be like foo(... true), foo(... false), etc. It might not as clear as what we have now.

Reply:
We could leave this for future work, it doesn't have to be done in this PR. It's a pre-existing problem and we don't have to block this PR on this.

> And on the other hand, some of them test both assertIsInstance and assertEqual, some of them test assertIsInstance only. So if you consolidate them into one function, you will have some kind of flag indicate whether to test assertEqual, and the function call will be like foo(... true), foo(... false), etc. It might not as clear as what we have now.

Actually, we could always do the isinstance and assertEqual checks. What those checks are checking is that (1) the Tensor has the correct dtype and (2) the tensor is on the correct device. For example,
- a lot of assertEqual(p.get_device()) calls are missing for cases where the input is on CPU. We can just assertEqual(p.get_device(), -1) for those.

Reply:
Oh, I see, you did this already. Thanks for improving the code!